pypit/artrace.py

```diff
@@ -1613,102 +1651,1643 @@ def slit_trace_qa(slf, frame, ltrace, rtrace, extslit, desc="",
+                    labelbottom='off', labelleft='off')
+
+    # Traces
+    iy = int(frame.shape[0]/2.)
+    for ii in range(ntrc):
+        if extslit[ii] is True:
+            ptyp = ':'
+        else:
+            ptyp = '--'
+        # Left
+        plt.plot(ltrace[:, ii]+0.5, ycen, 'r'+ptyp, linewidth=0.3, alpha=0.7)
+        # Right
+        plt.plot(rtrace[:, ii]+0.5, ycen, 'c'+ptyp, linewidth=0.3, alpha=0.7)
+        # Label
+        if use_slitid:
+            slitid, _, _ = arspecobj.get_slitid(slf, use_slitid, ii, ypos=0.5)
+            lbl = 'S{:04d}'.format(slitid)
+        else:
+            lbl = '{0:d}'.format(ii+1)
+        plt.text(0.5*(ltrace[iy, ii]+rtrace[iy, ii]), ycen[iy], lbl, color='green', ha='center', size='small')
+    # Title
+    tstamp = arqa.gen_timestamp()
+    if desc == "":
+        plt.suptitle(tstamp)
+    else:
+        plt.suptitle(desc+'\n'+tstamp)
+
+    # Write
+    plt.savefig(outfile, dpi=800)
+    plt.close()
+
+    plt.rcdefaults()
+
+
+def driver_trace_slits(mstrace, pixlocn, det=None, settings=None,
+                       ednum=100000, binbpx=None, ignore_orders=False,
+                       add_user_slits=None, armlsd=True):
+    """ Main driver for tracing slits.
+    May be renamed to trace_slits
+
+      Code flow
+       1.  Determine approximate slit edges (left, right)
+         1b.    Trim down to one pixel per edge per row [seems wasteful, but ok]
+       2.  Give edges ID numbers + stitch together partial edges (match_edges)
+         2b.   first maxgap option -- NOT recommended
+       3.  Assign slits (left, right) ::  Deep algorithm
+       4.  For ARMLSD
+          -- Trace crude the edges
+          -- Do a multi-slit sync to pair up left/right edges
+       5.  Remove short slits -- Not recommended for ARMLSD
+       6.  Fit left/right slits
+       7.  Synchronize
+       8.  Extrapolate into blank regions (PCA)
+
+    Parameters
+    ----------
+    mstrace : ndarray
+    pixlocn : ndarray
+    det : int (optional)
+    settings : dict (optional)
+    ednum : int (optional)
+       A large dummy number used for slit edge assignment.
+       ednum should be larger than the number of edges detected
+    binbpx : ndarray
+      Bad pixel mask
+    det : int (optional)
+      Required for single slilt specification
+    ignore_orders : bool (optional)
+      Perform ignore_orders algorithm (recommended only for echelle data)
+    add_user_slits : list of lists
+      List of 2 element lists, each an [xleft, xright] pair specifying a slit edge
+      These are specified at mstrace.shape[0]//2
+    armlsd : bool (optional)
+      Running with ARMLSD ?
+
+    Returns
+    -------
+    lcen : ndarray
+      Left edge traces
+    rcen  : ndarray
+      Right edge traces
+    extrapord
+    """
+    # Bad pixel array
+    if binbpx is None:
+        binbpx = np.zeros_like(mstrace)
+
+    if settings is None:  # Out of the function call because it is mutable
```

Comment:
What do you mean by the comment?  Still a bit confused about mutable arguments...

Reply:
discussed and resolved.